secret_hitler/game.py

```diff
@@ -49,55 +78,177 @@ def __init__(self, channel_id, game_id, max_players, admin_id):
+        random.shuffle(self.play_order)
 
         # Shuffle roles
         roles = config.configuration[self.max_players]["roles"]
         random.shuffle(roles)
 
-        for i in range(self.max_players):
-            self.players[i].role = roles[i]
+        for index, player in enumerate(self.players.values()):
+            player.role = roles[index]
 
-        self.president = self.players[0]
-        return True
-
-    def nominate(self, player_id):
+    def nominate(self, player_id, ctx=None):
         player = self.get_player(player_id)
+        if self.__is_player_eligible_for_nomination(player_id):
+            self.nominated = player
+            self.__transition_to_state(GameState.ELECTION, ctx=ctx)
+            return True
+        else:
+            return False
+
+    def __is_player_eligible_for_nomination(self, player_id: str) -> bool:
+        # Eligibility rules:
+        # 1. The last elected president and chancellor are ineligible to be chancellor.
+        # 2. If there are only five players left in the game, the last elected president
+        #    becomes eligible to be chancellor.
+        # 3. TODO: Veto power and the election tracker state can affect eligibility
+
         player_num = len(self.players)
         if player_num <= 5:
-            print(self.prev_chancellor_id)
-            print(player_id)
-            if self.chancellor is not None and self.chancellor.player_id is player_id:
+            if self.__is_player_chancellor(player_id):
                 return False
         elif player_num > 5:
-            if (self.chancellor is not None and self.chancellor.player_id is player_id) or self.prev_president_id is player_id:
+            if self.__is_player_chancellor(player_id) or self.__is_player_previous_president(player_id):
                 return False
 
-        if self.president is not None and self.president.player_id is player_id:
+        if self.__is_player_president(player_id):
             return False
-        self.nominated = player
-        self.state = GameStates.ELECTION
+        # All eligibility tests passed. This candidate is eligible.
         return True
 
-    def vote(self, player_id, vote):
-        for id, s_v in self.votes.items():
-            if id == player_id:
+    def vote(self, player_id: str, vote: str) -> VoteResult:
+        if self.state != GameState.ELECTION:
+            # TODO: We could raise an exception here and rely on the app module to ignore voting if no election is in process
+            return VoteResult.NO_ELECTION
+
+        is_vote_valid = self.__count_vote(player_id, vote)
+
+        if not is_vote_valid:
+            return VoteResult.INVALID
+
+        # If all votes are in, process the result of the election
+        if len(self.votes) == len(self.players):
+            vote_passed = self.__calculate_votes()
+            next_state = self.__next_state_after_vote(vote_passed)
+            self.__transition_to_state(next_state)
+            if vote_passed:
```

Comment:
I don't think we need to return additional VoteResult types here...the call is when someone clicks an emoji, either that click is valid or not.  Yes, it is also going to tally the votes once they are all in, but that info will get reported to the UI (app) via the callback mechanism during the state transition.  The caller has no reason to care what the outcome of the vote is in this context, it only cares if its vote was or was not valid.

Reply:
I wrote the VoteResult enum early on in the refactor and did not revisit it, but yeah, at this point, all that matters is indication of if the vote was counted or not.